### Table of contents

#### 1. Importing libraries

#### 2. Importing data

#### 3. Data consistency checks

#### 4. Tasks

#### 5. Exporting clean data

#### 1. Importing libraries

In [1]:
import pandas as pd
import numpy as np
import os

#### 2. Importing data

In [2]:
path = r"C:\Users\tseki\Documents\Learning\Careerfoundry\Data immersion\Achievement 4\Instacart Basket Analysis 8-6-22"

In [3]:
df_prods = pd.read_csv(os.path.join(path, 'Data', 'Original data', 'products.csv'), index_col = False)

In [4]:
# importing with index_col=false gives me an unmaed column which is annoying
df_ords = pd.read_csv(os.path.join(path, 'Data','Prepared data', 'orders_wrangled.csv'), index_col = False)

In [5]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [6]:
# I tried importing with index_col=[0] and it worked. I'm happy!
df_ords = pd.read_csv(os.path.join(path, 'Data','Prepared data', 'orders_wrangled.csv'), index_col = [0])

In [7]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

#### 3. Data consistency checks

In [8]:
# create a dataframe
df_test = pd.DataFrame()

In [9]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [16]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [17]:
# Checking column with mixed data type
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [18]:
df_test['mix'] = df_test['mix'].astype('str')

In [19]:
# Checking missed values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [20]:
# Creating sub-set with missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [21]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [22]:
df_prods.shape

(49693, 5)

In [23]:
# Creating subset without the missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [24]:
df_prods_clean.shape

(49677, 5)

In [ ]:
# Other ways of handling missing values includes dropping missing data; the code below can acheive this

# To drop all missing data: df_prods.dropna(inplace = True)
#To drop only the NaNs from a particular column: df_prods.dropna(subset = [‘product_name’], inplace = True)


In [25]:
# Finding duplicate rows
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [26]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [27]:
df_prods_clean.shape

(49677, 5)

In [28]:
# deleting duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [29]:
df_prods_clean_no_dups.shape

(49672, 5)

### Task

In [30]:
# Question 2 checking for possible errors in the data using the describe function
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

#### Ans: It can be observed that the maximum price is 99,999. This price seems rather outrageous.

In [21]:
# Question 3 checking for mixed type data in the order dataframe
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

#### Ans: there is no column with mixed data type in the order dataframe

In [22]:
# Question 5 checking for missed values in the order dataframe
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

#### Ans: the large number of missing values reported in the 'days_since_prior_order' column is as a result of the first order (represented by '1' in the 'order_number' column) placed by each customer. This is because 'days_since_prior_order' does not apply to the first order since they are first time customers.

In [23]:
# Question 6 Addressing missing values
# 1. create new dataframe
df_ord_clean = df_ords

In [24]:
# 2. create new column in the new dataframe
df_ord_clean['new_customer'] = df_ords['days_since_prior_order'].isnull() == True

#### I created a dataframe with a new column that identifies new customers with the boolean value 'true' and old customers with 'False'

In [25]:
df_ord_clean

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
1         2398795        1             2                   3   
2          473747        1             3                   3   
3         2254736        1             4                   4   
4          431534        1             5                   4   
...           ...      ...           ...                 ...   
3421078   2266710   206209            10                   5   
3421079   1854736   206209            11                   4   
3421080    626363   206209            12                   1   
3421081   2977660   206209            13                   1   
3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  new_customer  
0                        8                     NaN          True  
1                        7                    15.0         False  
2                       12                    21.0         False  
3                        7                    29.0         False  
4                       15                    28.0         False  
...                    ...                     ...           ...  
3421078                 18                    29.0         False  
3421079                 10                    30.0         False  
3421080                 12                    18.0         False  
3421081                 12                     7.0         False  
3421082                 14                    30.0         False  

[3421083 rows x 7 columns]

In [33]:
df_ord_clean.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  new_customer  
0                     NaN          True  
1                    15.0         False  
2                    21.0         False  
3                    29.0         False  
4                    28.0         False

In [26]:
# Checking that the number of rows with order number 1 is same is same as the sum of misssing values in the 'days_since_prior_order' column
df_check = df_ord_clean[df_ord_clean['order_number']==1]

In [27]:
df_check.shape

(206209, 7)

In [28]:
# Question 7 Finding duplicate data in the order data frame
df_dups_ord = df_ord_clean[df_ord_clean.duplicated()]

In [29]:
df_dups_ord

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, new_customer]
Index: []

#### Ans: There are no duplicates in the order dataframe

### Exporting clean data

In [50]:
# Exporting product data
df_prods_clean_no_dups.to_csv(os.path.join(path, 'Data', 'Prepared data', 'products_clean.csv'))

In [36]:
# Exporting order data
df_ord_clean.to_csv(os.path.join(path, 'Data', 'Prepared Data', 'orders_clean.csv'))